[vulkan] Efficient gemm implementation

SS-JIA opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49609 [vulkan] Efficient gemm implementation**

Differential Revision: [D26209677](https://our.internmc.facebook.com/intern/diff/D26209677)

Commit:
[vulkan] Efficient gemm implementation

[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"


Differential Revision: [D26209677](https://our.internmc.facebook.com/intern/diff/D26209677)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Mm.cpp:
```diff
@@ -281,78 +276,136 @@ Tensor LinearOpContext::run(
 
+  api::Command::Buffer& output_unpack_buffer = command_pool.stream();
+  vTensor v_output = unpack_image2d_h2w2(v_output_packed, output_sizes, context, output_unpack_buffer);
+  command_pool.submit(context->gpu().queue, output_unpack_buffer);
```
Same.  Pack / unpack can go inside same submission.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Mm.h:
```diff
@@ -32,11 +32,13 @@ class LinearOpContext final : public torch::jit::CustomClassHolder {
   struct {
     Tensor weight;
     c10::optional<Tensor> bias;
+    bool has_bias;
```
Can you use bias && bias->defined() in place of both has_bias variables?

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Packing.cpp:
```diff
@@ -0,0 +1,160 @@
+
+using namespace api::utils;
+
+vTensor pack_image2d_h2w2(vTensor v_src, api::Context* context, api::Command::Buffer& command_buffer) {
```
const correctness please.  Pass vTensor by const ref, and change `api::Context* context` to `api::Context* const context` to mark the pointer (not the pointee) immutable.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Packing.cpp:
```diff
@@ -0,0 +1,160 @@
+    },
+    v_src.options()
+  };
+  const struct {
```
Same comment regarding anonymous structs on GCC.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Packing.cpp:
```diff
@@ -0,0 +1,160 @@
+  };
+
+  uint32_t orig_w = output_sizes[output_sizes.size() - 1];
+  uint32_t orig_h = output_sizes[output_sizes.size() - 2];
```
const.  const everywhere please. I am a const zealot. :)

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Packing.cpp:
```diff
@@ -0,0 +1,160 @@
+  return v_src_unpacked;
+}
+
+vTensor unpack_image1x1(vTensor v_src, c10::SmallVector<int64_t, 4u> output_sizes, api::Context* context, api::Command::Buffer& command_buffer) {
```
Pass all objects greater than the size of the two machine words (2 x 64-bits on 64-bit, 2 x 32-bit for 32-bits) by [const] reference.  I add a fudge factor of 2 since pointer chasing and dereferencing (which is effectively what references are - just syntactic sugar for pointers) has a cost so it's best avoided when the cost of passing by value is small.

Commit:
Update on "[vulkan] Efficient gemm implementation"


Differential Revision: [D26209677](https://our.internmc.facebook.com/intern/diff/D26209677)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Packing.h:
```diff
@@ -0,0 +1,18 @@
+vTensor pack_image2d_h2w2(vTensor v_src, api::Context* context, api::Command::Buffer& command_buffer);
+vTensor unpack_image2d_h2w2(vTensor v_src, c10::SmallVector<int64_t, 4u> output_sizes, api::Context* context, api::Command::Buffer& command_buffer);
+
+vTensor unpack_image1x1(vTensor v_src, c10::SmallVector<int64_t, 4u> output_sizes, api::Context* context, api::Command::Buffer& command_buffer);
```
If these functions are only used in one single implementation file, please remove the common header.  Reason: Software engineering is the art (since it is not all science unfortunately) and science of change management, and the bedrock of managing changes is limiting scope.  Limiting scope in general is the single most important tool software engineers have to get a handle on entropy.

AshkanAliabadi commented on aten/src/ATen/test/vulkan_api_test.cpp:
```diff
@@ -156,15 +208,12 @@ TEST(VulkanAPITest, addmm) {
   if (!check) {
-    std::cout << "Expected:\n" << out_cpu << std::endl;
-    std::cout << "Got:\n" << out_vulkan.cpu() << std::endl;
+    showRtol(out_cpu, out_vulkan.cpu());
```
Change other places to this function as well.

AshkanAliabadi reviewed (approved):
Thanks Stephen!

SS-JIA commented on aten/src/ATen/native/vulkan/glsl/mean2d.glsl:
```diff
@@ -12,29 +12,42 @@ layout(set = 0, binding = 2) uniform PRECISION restrict           Block {
+  const ivec3 group_size = ivec3(gl_WorkGroupSize);
 
-  if (all(lessThan(pos, uBlock.size.xyz))) {
+  if (pos.z < uBlock.size.z) {
```
It's not needed since the loop below is constrained by the size of the input tensor.

SS-JIA commented on aten/src/ATen/native/vulkan/glsl/addmm.glsl:
```diff
@@ -23,10 +23,10 @@ void main() {
+      vec4 texel1 = texelFetch(uM1, ivec3(k, pos.y, pos.z), 0);
+      vec4 texel2 = texelFetch(uM2, ivec3(pos.x, k, pos.z), 0);
+      sum = fma(texel1.xxzz, texel2.xyxy, sum);
+      sum = fma(texel1.yyww, texel2.zwzw, sum);
```
Yes, the new packing makes use of the entire input texel.

Commit:
Update on "[vulkan] Efficient gemm implementation"


Differential Revision: [D26209677](https://our.internmc.facebook.com/intern/diff/D26209677)

[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"


Differential Revision: [D26209677](https://our.internmc.facebook.com/intern/diff/D26209677)

[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"


Differential Revision: [D26209677](https://our.internmc.facebook.com/intern/diff/D26209677)

[ghstack-poisoned]

Commit:
Update on "[vulkan] Efficient gemm implementation"


Differential Revision: [D26209677](https://our.internmc.facebook.com/intern/diff/D26209677)

[ghstack-poisoned]
